File: src/orderIntentLocation.js
import { deriveLocalityKey, parseGeoCoord } from "./donorNeighbourhoodArea.js";

/**
 * @param {unknown} payload
 * @returns {{ lat: number, lng: number, label: string, localityKey: string } | null}
 */
export function locationFromPayload(payload) {
  if (!payload || typeof payload !== "object") {
    return null;
  }
  const lat = parseGeoCoord(payload.location_lat, "lat");
  const lng = parseGeoCoord(payload.location_lng, "lng");
  if (lat == null || lng == null) {
    return null;
  }
  const label =
    typeof payload.location_label === "string"
      ? payload.location_label.trim()
      : "";
  const suppliedKey =
    typeof payload.locality_key === "string"
      ? payload.locality_key.trim()
      : "";
  const localityKey = suppliedKey || deriveLocalityKey(lat, lng);
  return { lat, lng, label, localityKey };
}

/**
 * @param {object} record
 * @param {ReturnType<typeof locationFromPayload>} location
 */
export function applyLocationToRecord(record, location) {
  if (!location) {
    return {
      ...record,
      location_lat: null,
      location_lng: null,
      location_label: "",
      locality_key: ""
    };
  }
  return {
    ...record,
    location_lat: location.lat,
    location_lng: location.lng,
    location_label: location.label,
    locality_key: location.localityKey
  };
}

export function mergeLocationFromPayload(existing, payload) {
  const incoming = locationFromPayload(payload);
  if (incoming) {
    return applyLocationToRecord(existing, incoming);
  }
  if (payload?.location_lat === null && payload?.location_lng === null) {
    return applyLocationToRecord(existing, null);
  }
  return existing;
}

export function recordHasLocation(record) {
  return (
    typeof record?.location_lat === "number" &&
    typeof record?.location_lng === "number" &&
    Number.isFinite(record.location_lat) &&
    Number.isFinite(record.location_lng)
  );
}

File: src/main/java/org/sharingbridge/integration/service/OrderIntentLocation.java
package org.sharingbridge.integration.service;

import java.util.LinkedHashMap;
import java.util.Map;
import org.sharingbridge.integration.geo.DonorNeighbourhoodArea;
import org.sharingbridge.integration.geo.LocalityKey;

public final class OrderIntentLocation {

    private OrderIntentLocation() {}

    public record Location(double lat, double lng, String label, String localityKey) {}

    public static Location locationFromPayload(
            Map<String, Object> payload, PostalGeocodeService geocode) {
        if (payload == null) {
            return null;
        }
        Double lat = DonorNeighbourhoodArea.parseGeoCoord(payload.get("location_lat"), "lat");
        Double lng = DonorNeighbourhoodArea.parseGeoCoord(payload.get("location_lng"), "lng");
        if (lat == null || lng == null) {
            return null;
        }
        String label = payload.get("location_label") instanceof String text ? text.trim() : "";
        String suppliedKey =
                payload.get("locality_key") instanceof String text ? text.trim() : "";
        String localityKey = "";
        if (!suppliedKey.isEmpty() && LocalityKey.isValidLocalityKey(suppliedKey)) {
            localityKey = suppliedKey;
        } else if (geocode != null) {
            String derived = geocode.derivePostalLocalityKey(lat, lng);
            localityKey = derived == null ? "" : derived;
        }
        return new Location(lat, lng, label, localityKey);
    }

    public static Map<String, Object> applyLocationToRecord(
            Map<String, Object> record, Location location) {
        Map<String, Object> next = new LinkedHashMap<>(record);
        if (location == null) {
            next.put("location_lat", null);
            next.put("location_lng", null);
            next.put("location_label", "");
            next.put("locality_key", "");
            return next;
        }
        next.put("location_lat", location.lat());
        next.put("location_lng", location.lng());
        next.put("location_label", location.label());
        next.put("locality_key", location.localityKey());
        return next;
    }

    public static Map<String, Object> mergeLocationFromPayload(
            Map<String, Object> existing, Map<String, Object> payload, PostalGeocodeService geocode) {
        Location incoming = locationFromPayload(payload, geocode);
        if (incoming != null) {
            return applyLocationToRecord(existing, incoming);
        }
        if (payload != null
                && payload.get("location_lat") == null
                && payload.containsKey("location_lat")
                && payload.get("location_lng") == null
                && payload.containsKey("location_lng")) {
            return applyLocationToRecord(existing, null);
        }
        return existing;
    }

    public static boolean recordHasLocation(Map<String, Object> record) {
        if (record == null) {
            return false;
        }
        Object lat = record.get("location_lat");
        Object lng = record.get("location_lng");
        return lat instanceof Number
                && lng instanceof Number
                && Double.isFinite(((Number) lat).doubleValue())
                && Double.isFinite(((Number) lng).doubleValue());
    }
}
